### PR TITLE
[FIX] Remove non-steady-state volumes prior to ICA-AROMA

### PR DESCRIPTION
**Pull request by jdkent**
## Changes proposed in this pull request
ref #1136
ref #1300

This pull request pulls the steady state detector into `bold/base.py` and uses it's output for both the confounds workflow and the aroma workflow.

## Documentation that should be reviewed

I need to update the documentation for ica-aroma, more details to come...

### CODE REVIEW

File: fmriprep/workflows/bold/base.py [hunk 1]
```diff
@@ -424,6 +425,9 @@ def init_func_preproc_wf(bold_file, ignore, freesurfer,
                                              omp_nthreads=omp_nthreads,
                                              use_compression=False)
 
+    # get non steady-state volumes
+    non_steady_state = pe.Node(nac.NonSteadyStateDetector(), name='non_steady_state')
```
Comment (effigies) [inline fmriprep/workflows/bold/base.py]: You should already be able to get this from `bold_reference_wf.outputnode.skip_vols`.

Comment (jdkent) [reply fmriprep/workflows/bold/base.py]: nice, thanks!

File: fmriprep/workflows/bold/confounds.py [hunk 2]
```diff
@@ -586,18 +612,67 @@ def _getbtthresh(medianval):
...
+
+        # modify header with new shape (fewer volumes)
+        data_shape = list(bold_img.header.get_data_shape())
+        data_shape[-1] -= skip_vols
+        bold_img.header.set_data_shape(tuple(data_shape))
+
+        # save the resulting bold file
+        out = fname_presuffix(bold_file, suffix='_cut')
+        bold_img.__class__(bold_data_cut, bold_img.affine, bold_img.header).to_filename(out)
+        return out
```
Comment (effigies) [inline fmriprep/workflows/bold/confounds.py]: This could be much shorter:

```Python
import nibabel as nb
from nipype.utils.filemanip import fname_presuffix

out = fname_presuffix(bold_file, suffix='_cut')
bold_img = nb.load(bold_file)
bold_img.__class__(bold_img.dataobj[..., skip_vols:],
                   bold_img.affine, bold_img.header).to_filename(out)
return out
```

Note that the data shape will be set automatically, so there's no need to manually update it. And using the `dataobj` should save as much memory as you can, though I don't think there's much waste here, assuming `skip_vols` is small.

Or, with nibabel>=2.3:

```Python
import nibabel as nb
from nipype.utils.filemanip import fname_presuffix

out = fname_presuffix(bold_file, suffix='_cut')
nb.load(bold_file).slicer[..., skip_vols:].to_filename(out)
return out
```

**Comment by jdkent** [comment]
If I wanted to write tests for the semi-private functions, where should I put them? Should I test them?

File: fmriprep/workflows/bold/confounds.py [hunk 3]
```diff
@@ -586,18 +612,67 @@ def _getbtthresh(medianval):
...
+    from nipype.utils.filemanip import fname_presuffix
+
+    # load the data
+    bold_img = nb.load(bold_file)
+    bold_data = bold_img.get_data()
+    bold_cut_img = nb.load(bold_cut_file)
+    bold_cut_data = bold_cut_img.get_data()
+
+    # assign everything from skip_vols foward to bold_cut_data
+    bold_data[..., skip_vols:] = bold_cut_data
```
Comment (effigies) [inline fmriprep/workflows/bold/confounds.py]: What about:

```Python
bold_img = nb.load(bold_file)
bold_cut_img = nb.load(bold_cut_file)

bold_data = np.concatenate((bold_img.dataobj[..., :skip_vols],
                            bold_cut_img.dataobj), axis=3)
```

This would reduce the memory usage by the uncompressed size of `bold_cut_img`.

Or, if we're willing to require `nibabel>=2.3`:

```Python
new_img = nb.concat_images((bold_img.slicer[..., :skip_vols], bold_cut_img), axis=3)
```

This last has the advantage of checking affines to ensure the images are in the same space.

File: fmriprep/workflows/bold/confounds.py [hunk 4]
```diff
@@ -586,18 +612,67 @@ def _getbtthresh(medianval):
...
     ])
 
     return workflow
 
 
+def _remove_volumes(bold_file, skip_vols):
+        import nibabel as nb
+        from nipype.utils.filemanip import fname_presuffix
+
+        # load the bold file and get the 4d matrix
```
Comment (effigies) [inline fmriprep/workflows/bold/confounds.py]: Before doing anything else, I would check:

```Python
if skip_vols == 0:
    return bold_file
```

File: fmriprep/workflows/bold/confounds.py [hunk 5]
```diff
@@ -586,18 +612,67 @@ def _getbtthresh(medianval):
...
+        bold_img.__class__(bold_data_cut, bold_img.affine, bold_img.header).to_filename(out)
+        return out
+
+
+def _add_volumes(bold_file, bold_cut_file, skip_vols):
+    """prepend skip_vols from bold_file onto bold_cut_file"""
+    import nibabel as nb
+    from nipype.utils.filemanip import fname_presuffix
+
+    # load the data
```
Comment (effigies) [inline fmriprep/workflows/bold/confounds.py]:
```Python
if skip_vols == 0:
    return bold_cut_file
```

**Comment by effigies** [comment]
@jdkent We're trying to update to be largely conformant with BIDS Derivatives RC2 (due out Friday-ish). That will trigger fMRIPrep 1.2.0. Do you want to get this in by then, or do you need more time?

**Comment by jdkent** [comment]
Hi @effigies, Thanks for the helpful review! I've integrated the changes you suggested in the last commit, but I have not written tests yet. I think I will be able generate tests by friday, but if not, I can make it a separate pull request, and have this one merged as is (if all the tests pass).

File: fmriprep/workflows/bold/confounds.py [hunk 6]
```diff
@@ -428,40 +427,53 @@ def init_ica_aroma_wf(template, metadata, mem_gb, omp_nthreads,
...
         susan_fwhm : float
             Kernel width (FWHM in mm) for the smoothing step with
             FSL ``susan`` (default: 6.0mm)
         use_fieldwarp : bool
             Include SDC warp in single-shot transform from BOLD to MNI
-        ignore_aroma_err : bool
-            Do not fail on ICA-AROMA errors
-        aroma_melodic_dim: int or None
-            Set the dimensionality of the Melodic ICA decomposition
-            If None, MELODIC automatically estimates dimensionality.
```
Comment (effigies) [inline fmriprep/workflows/bold/confounds.py]: It looks like you sorted these. I would keep these in the order that they are in the parameter list. If we want to reorder that, we can do that in another PR, but that's an API change, and we should avoid it unless there's a good reason.

**Comment by effigies** [comment]
@jdkent I'm going to go ahead and merge. Please submit tests in a new PR.
